lib/Compilation.js

```diff
@@ -35,153 +35,319 @@ const createHash = require("./util/createHash");
+ * @property {Dependency[]} dependencies
+ */
+
+/**
+ * @typedef {Object} ModuleFactory
+ * @property {(data: ModuleFactoryCreateData, callback: ModuleCallback) => any} create
+ */
+
+/**
+ * @typedef {Object} SortedDependency
+ * @property {ModuleFactory} factory
+ * @property {Dependency[]} dependencies
+ */
+
+/**
+ * @typedef {Object} AvailableModulesChunkGroupMapping
+ * @property {ChunkGroup} chunkGroup
+ * @property {Set<Module>} availableModules
+ */
+
+/**
+ * @typedef {Object} DependenciesBlockLike
+ * @property {Dependency[]} dependencies
+ * @property {AsyncDependenciesBlock[]} blocks
+ * @property {DependenciesBlockVariable[]} variables
+ */
+
+/**
+ * @param {Chunk} a first chunk to sort by id
+ * @param {Chunk} b second chunk to sort by id
+ * @returns {-1|0|1} sort value
+ */
+function byId(a, b) {
+	if (a.id && b.id) {
+		if (a.id < b.id) return -1;
+		if (a.id > b.id) return 1;
+	}
 	return 0;
-};
+}
 
-const byIdOrIdentifier = (a, b) => {
+/**
+ * @param {Module} a first module to sort by
+ * @param {Module} b second module to sort by
+ * @returns {-1|0|1} sort value
+ */
+function byIdOrIdentifier(a, b) {
 	if (a.id < b.id) return -1;
 	if (a.id > b.id) return 1;
 	const identA = a.identifier();
 	const identB = b.identifier();
 	if (identA < identB) return -1;
 	if (identA > identB) return 1;
 	return 0;
-};
+}
 
-const byIndexOrIdentifier = (a, b) => {
+/**
+ * @param {Module} a first module to sort by
+ * @param {Module} b second module to sort by
+ * @returns {-1|0|1} sort value
+ */
+function byIndexOrIdentifier(a, b) {
 	if (a.index < b.index) return -1;
 	if (a.index > b.index) return 1;
 	const identA = a.identifier();
 	const identB = b.identifier();
 	if (identA < identB) return -1;
 	if (identA > identB) return 1;
 	return 0;
-};
+}
 
-const byNameOrHash = (a, b) => {
+/**
+ * @param {Compilation} a first compilation to sort by
+ * @param {Compilation} b second compilation to sort by
+ * @returns {-1|0|1} sort value
+ */
+function byNameOrHash(a, b) {
 	if (a.name < b.name) return -1;
 	if (a.name > b.name) return 1;
 	if (a.fullHash < b.fullHash) return -1;
-	if (a.fullhash > b.fullHash) return 1;
+	if (a.fullHash > b.fullHash) return 1;
 	return 0;
-};
+}
 
-const iterationBlockVariable = (variables, fn) => {
+/**
+ * @param {DependenciesBlockVariable[]} variables DepBlock Variables to iterate over
+ * @param {DepBlockVarDependenciesCallback} fn callback to apply on iterated elements
+ * @returns {void}
+ */
+function iterationBlockVariable(variables, fn) {
 	for (
 		let indexVariable = 0;
 		indexVariable < variables.length;
 		indexVariable++
 	) {
+		// TODO: Remove when Dependency + DependenciesBlockVariable is typed
+		/** @type {Dependency[]} */
 		const varDep = variables[indexVariable].dependencies;
 		for (let indexVDep = 0; indexVDep < varDep.length; indexVDep++) {
 			fn(varDep[indexVDep]);
 		}
 	}
-};
+}
 
-const iterationOfArrayCallback = (arr, fn) => {
+/**
+ * @template {T}
```

Comment:
although Typescript currently allows surrounding the name of the type variable with `{}`, it's not standard and may break when we introduce type parameter constraints, which have the proposed syntax `/** @template {Constraint} T */` (equivalent to Typescript's `T extends Constraint`)